clang/lib/AST/ExprConstant.cpp

```diff
@@ -13866,6 +13940,157 @@ bool FixedPointExprEvaluator::VisitBinaryOperator(const BinaryOperator *E) {
   return Success(Result, E);
 }
 
+bool DecimalFloatExprEvaluator::VisitUnaryOperator(const UnaryOperator *E) {
+  switch (E->getOpcode()) {
+    default:
+      // Invalid unary operators
+      return Error(E);
+    case UO_Plus:
+      // The result is just the value.
+      return Visit(E->getSubExpr());
+    case UO_Minus: {
+      if (!Visit(E->getSubExpr())) return false;
+      if (!Result.isDecimalFloat())
+        return Error(E);
+      bool Overflowed;
+      APDecimalFloat Negated = Result.getDecimalFloat().negate(&Overflowed);
+      if (Overflowed && !HandleOverflow(Info, E, Negated, E->getType()))
+        return false;
+      return Success(Negated, E);
+    }
+  }
+}
+
+bool DecimalFloatExprEvaluator::VisitBinaryOperator(const BinaryOperator *E) {
+  if (E->isPtrMemOp() || E->isAssignmentOp() || E->getOpcode() == BO_Comma)
+    return ExprEvaluatorBaseTy::VisitBinaryOperator(E);
+
+  const Expr *LHS = E->getLHS();
+  const Expr *RHS = E->getRHS();
+  llvm::decFltSemantics ResultFXSema =
+      Info.Ctx.getDecimalFloatSemantics(E->getType());
+
+  APDecimalFloat LHSFX(Info.Ctx.getDecimalFloatSemantics(LHS->getType()));
+  if (!EvaluateDecimalFloat(LHS, LHSFX, Info))
+    return false;
+  APDecimalFloat RHSFX(Info.Ctx.getDecimalFloatSemantics(RHS->getType()));
+  if (!EvaluateDecimalFloat(RHS, RHSFX, Info))
+    return false;
+
+  bool OpOverflow = false, ConversionOverflow = false;
+  APDecimalFloat Result(LHSFX.getSemantics());
+  switch (E->getOpcode()) {
+  case BO_Add: {
+    Result = LHSFX.add(RHSFX, &OpOverflow)
+                  .convert(ResultFXSema, &ConversionOverflow);
+    break;
+  }
+  case BO_Sub: {
+    Result = LHSFX.sub(RHSFX, &OpOverflow)
+                  .convert(ResultFXSema, &ConversionOverflow);
+    break;
+  }
+  case BO_Mul: {
+    Result = LHSFX.mul(RHSFX, &OpOverflow)
+                  .convert(ResultFXSema, &ConversionOverflow);
+    break;
+  }
+  case BO_Div: {
+    if (RHSFX.isZero()) {
+      Info.FFDiag(E, diag::note_expr_divide_by_zero);
+      return false;
+    }
+    Result = LHSFX.div(RHSFX, &OpOverflow)
+                  .convert(ResultFXSema, &ConversionOverflow);
+    break;
+  }
+  
+  default:
+    return false;
+  }
+  if (OpOverflow || ConversionOverflow) {
+    if (Info.checkingForUndefinedBehavior())
+      Info.Ctx.getDiagnostics().Report(E->getExprLoc(),
+                                       diag::warn_fixedpoint_constant_overflow)
```

Comment:
The referenced diagnostic is for fixed-point, so we presumably need to use the right one (if one for DFP already exists), introduce a new one or, if the existing one is sufficiently generic, rename it. Additional occurrences can be found below.